README.md
Prooph\Link\MessageQueue
========================
Message Queue Module for [prooph LINK](https://github.com/prooph/link)

# Bernard Backend
This module provides a simple message queue for prooph LINK, so that workflow messages can be handled asynchronous.
The message queue is implemented with the help of [bernard](https://github.com/bernardphp/bernard) and the corresponding
[message dispatcher](https://github.com/prooph/psb-bernard-dispatcher) implementation for prooph service bus.
It uses the DoctrineDriver shipped with bernard and connects to the `prooph.link.app.db` database connection to manage the
message queue.

# Message Dispatcher
The module provides a ready to use message dispatcher that can be defined in a processing channel to push workflow messages
on the queue send over the channel. A channel config using the message dispatcher would look something like this:

```php
return [
    'processing' => [
        'channels' => [
            'message_queue' => [
                'targets' => ['Acme\WorkflowMessageHandler'],
                'sender'  => 'a_node_name',
                 'message_dispatcher' => 'prooph.link.message_queue.producer' //ServiceManager alias of the message dispatcher
            ],
            //...
        ],
        //...
    ],
    //...
];
```

# Workflow Processor Message Queue
The [app-core](https://github.com/prooph/link-app-core) module provides the possibility to activate a general message queue for
all messages sent by the workflow processor. The message-queue module provides the required implementation. It defines the
`\Prooph\Link\Application\Definition::APP_SERVICE_WORKFLOW_PROCESSOR_MESSAGE_QUEUE` as an alias of the `prooph.link.message_queue.producer`
so the message dispatcher handles all messages sent by the workflow processor.

# Message Consumer
To consume messages pushed to the queue a TickOccurred event needs to be published by the application. You can read about it in the
app-core module documentation. The message consumer shipped with the message-queue module listens on this TickOccurred event, performs
a bernard consumer tick (pulls the next pending message from the queue) and forwards the message to the processing workflow engine without
further routing. The workflow engine is responsible for routing the message to its target.

# Support

- Ask any questions on [prooph-users](https://groups.google.com/forum/?hl=de#!forum/prooph) google group.
- File issues at [https://github.com/prooph/link-app-core/issues](https://github.com/prooph/link-message-queue/issues).

# Contribution

You wanna help us? Great!
We appreciate any help, be it on implementation level, UI improvements, testing, donation or simply trying out the system and give us feedback.
Just leave us a note in our google group linked above and we can discuss further steps.

Thanks,
your prooph team


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]

## [0.1.0] - 2015-05-10
### Added
- First development version of the module


[unreleased]: https://github.com/prooph/link-message-queue/compare/v0.1.0...HEAD

config/module.config.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/29/15 - 10:42 PM
 */
return [
    'prooph.link.message_queue' => [
        'queue_name' => 'link_processing_queue',
        'db' => [
            //The module uses the prooph.link.app.db connection by default
            'use_application_db' => true,
            //If you set this flag to false you have to provide a doctrine dbal connection configuration
            //'connection' => [/* connection settings go here */]
        ],
    ],
    'service_manager' => [
        'factories' => [
            'prooph.link.message_queue.persistent_factory' => \Prooph\Link\MessageQueue\Producer\PersistentFactoryFactory::class,
            'prooph.link.message_queue.producer' => \Prooph\Link\MessageQueue\Producer\PsbBernardMessageDispatcherFactory::class,
            'prooph.link.message_queue.bernard_workflow_engine_router' => \Prooph\Link\MessageQueue\Consumer\BernardWorkflowEngineRouterFactory::class,
            'prooph.link.message_queue.consumer' => \Prooph\Link\MessageQueue\Consumer\MessageQueueConsumerFactory::class,

        ],
        'aliases' => [
            \Prooph\Link\Application\Definition::APP_SERVICE_WORKFLOW_PROCESSOR_MESSAGE_QUEUE => 'prooph.link.message_queue.producer',
        ]
    ],
    'proophessor' => [
        'event_router_map' => [
            \Prooph\Link\Application\Event\TickOccurred::class => [
                'prooph.link.message_queue.consumer'
            ]
        ]
    ],
];

src/Consumer/MessageQueueConsumer.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/30/15 - 12:00 AM
 */
namespace Prooph\Link\MessageQueue\Consumer;
use Bernard\Consumer;
use Bernard\Queue;
use Prooph\Link\Application\Event\TickOccurred;

/**
 * Class MessageQueueConsumer
 *
 * The message queue consumer listens on the Prooph\Link\Application\Event\TickOccurred event and triggers
 * the psb bernard message consumer to pull the next pending message from the message queue.
 *
 * @package Prooph\Link\MessageQueue\Consumer
 * @author Alexander Miertsch <contact@prooph.de>
 */
final class MessageQueueConsumer 
{
    /**
     * @var Consumer
     */
    private $psbBernardMessageConsumer;

    /**
     * @var Queue
     */
    private $queue;

    /**
     * @param Consumer $consumer
     * @param Queue $queue
     */
    public function __construct(Consumer $consumer, Queue $queue)
    {
        $this->psbBernardMessageConsumer = $consumer;
        $this->queue = $queue;
    }

    /**
     * @param TickOccurred $event
     */
    public function onTickOccurred(TickOccurred $event)
    {
        $this->psbBernardMessageConsumer->tick($this->queue);
    }
} 

src/Consumer/BernardWorkflowEngineRouter.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/30/15 - 12:05 AM
 */
namespace Prooph\Link\MessageQueue\Consumer;

use Bernard\Envelope;
use Bernard\Router;
use Prooph\Common\Messaging\RemoteMessage;
use Prooph\Processing\Processor\WorkflowEngine;
use Prooph\ServiceBus\Message\Bernard\BernardMessage;

final class BernardWorkflowEngineRouter implements Router
{
    /**
     * @var WorkflowEngine
     */
    private $workflowEngine;

    /**
     * @param WorkflowEngine $workflowEngine
     */
    public function __construct(WorkflowEngine $workflowEngine)
    {
        $this->workflowEngine = $workflowEngine;
    }

    /**
     * Returns the right Receiver (callable) based on the Envelope.
     *
     * @param  Envelope $envelope
     * @throws \InvalidArgumentException
     * @return array
     */
    public function map(Envelope $envelope)
    {
        $message = $envelope->getMessage();

        if (! $message instanceof BernardMessage) {
            throw new \InvalidArgumentException(sprintf(
                "Routing the message %s failed due to wrong message type",
                $envelope->getName()
            ));
        }

        return array($this, "routeMessage");
    }

    /**
     * @param BernardMessage $message
     */
    public function routeMessage(BernardMessage $message)
    {
        $this->workflowEngine->dispatch($message->getRemoteMessage(), 'prooph.link.message_queue.consumer');
    }
}

src/Consumer/MessageQueueConsumerFactory.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/30/15 - 12:17 AM
 */
namespace Prooph\Link\MessageQueue\Consumer;

use Bernard\Consumer;
use Bernard\Middleware\MiddlewareBuilder;
use Bernard\QueueFactory\PersistentFactory;
use Zend\ServiceManager\FactoryInterface;
use Zend\ServiceManager\ServiceLocatorInterface;

/**
 * Class MessageQueueConsumerFactory
 *
 * @package Prooph\Link\MessageQueue\Consumer
 * @author Alexander Miertsch <contact@prooph.de>
 */
final class MessageQueueConsumerFactory implements FactoryInterface
{

    /**
     * Create service
     *
     * @param ServiceLocatorInterface $serviceLocator
     * @throws \RuntimeException
     * @return mixed
     */
    public function createService(ServiceLocatorInterface $serviceLocator)
    {
        $config = $serviceLocator->get('config');

        if (! isset($config['prooph.link.message_queue'])) {
            throw new \RuntimeException('Missing prooph.link.message_queue key in application config.');
        }

        $config = $config['prooph.link.message_queue'];

        if (! isset($config['queue_name'])) {
            throw new \RuntimeException('Missing queue_name for prooph.link.message_queue in application config');
        }

        /** @var $persistentFactory PersistentFactory */
        $persistentFactory = $serviceLocator->get('prooph.link.message_queue.persistent_factory');

        return new MessageQueueConsumer(
            new Consumer($serviceLocator->get('prooph.link.message_queue.bernard_workflow_engine_router'), new MiddlewareBuilder()),
            $persistentFactory->create($config['queue_name'])
        );
    }
}

src/Consumer/BernardWorkflowEngineRouterFactory.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/30/15 - 12:14 AM
 */
namespace Prooph\Link\MessageQueue\Consumer;

use Prooph\Processing\Environment\Environment;
use Zend\ServiceManager\FactoryInterface;
use Zend\ServiceManager\ServiceLocatorInterface;

/**
 * Class BernardWorkflowEngineRouterFactory
 *
 * @package Prooph\Link\MessageQueue\Consumer
 * @author Alexander Miertsch <contact@prooph.de>
 */
final class BernardWorkflowEngineRouterFactory implements FactoryInterface
{

    /**
     * Create service
     *
     * @param ServiceLocatorInterface $serviceLocator
     * @return mixed
     */
    public function createService(ServiceLocatorInterface $serviceLocator)
    {
        /** @var $env Environment */
        $env = $serviceLocator->get('processing_environment');

        return new BernardWorkflowEngineRouter($env->getWorkflowEngine());
    }
}

src/Producer/PsbBernardMessageDispatcherFactory.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/29/15 - 11:30 PM
 */
namespace Prooph\Link\MessageQueue\Producer;

use Bernard\Driver\DoctrineDriver;
use Bernard\Middleware\MiddlewareBuilder;
use Bernard\Producer;
use Bernard\QueueFactory\PersistentFactory;
use Prooph\ServiceBus\Message\Bernard\BernardSerializer;
use Prooph\ServiceBus\Message\Bernard\MessageDispatcher;
use Zend\ServiceManager\FactoryInterface;
use Zend\ServiceManager\ServiceLocatorInterface;
use Doctrine\DBAL\DriverManager;

/**
 * Class PsbBernardMessageDispatcherFactory
 *
 * @package Prooph\Link\MessageQueue\Producer
 * @author Alexander Miertsch <contact@prooph.de>
 */
final class PsbBernardMessageDispatcherFactory implements FactoryInterface
{
    /**
     * Create service
     *
     * @param ServiceLocatorInterface $serviceLocator
     * @throws \RuntimeException
     * @return mixed
     */
    public function createService(ServiceLocatorInterface $serviceLocator)
    {
        $config = $serviceLocator->get('config');

        if (! isset($config['prooph.link.message_queue'])) {
            throw new \RuntimeException('Missing prooph.link.message_queue key in application config.');
        }

        $config = $config['prooph.link.message_queue'];

        if (! isset($config['queue_name'])) {
            throw new \RuntimeException('Missing queue_name for prooph.link.message_queue in application config');
        }

        $bernardProducer = new Producer($serviceLocator->get('prooph.link.message_queue.persistent_factory'), new MiddlewareBuilder());

        return new MessageDispatcher($bernardProducer, $config['queue_name']);
    }
}

src/Producer/PersistentFactoryFactory.php
<?php
/*
 * This file is part of prooph/link.
 * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * Date: 3/29/15 - 11:55 PM
 */
namespace Prooph\Link\MessageQueue\Producer;

use Bernard\Driver\DoctrineDriver;
use Bernard\QueueFactory\PersistentFactory;
use Doctrine\DBAL\DriverManager;
use Prooph\ServiceBus\Message\Bernard\BernardSerializer;
use Zend\ServiceManager\FactoryInterface;
use Zend\ServiceManager\ServiceLocatorInterface;

/**
 * Class PersistentFactoryFactory
 *
 * Creates a bernard persistent factory
 *
 * @package Prooph\Link\MessageQueue\Producer
 * @author Alexander Miertsch <contact@prooph.de>
 */
final class PersistentFactoryFactory implements FactoryInterface
{

    /**
     * Create service
     *
     * @param ServiceLocatorInterface $serviceLocator
     * @throws \RuntimeException
     * @return mixed
     */
    public function createService(ServiceLocatorInterface $serviceLocator)
    {
        $config = $serviceLocator->get('config');

        if (! isset($config['prooph.link.message_queue'])) {
            throw new \RuntimeException('Missing prooph.link.message_queue key in application config.');
        }

        $config = $config['prooph.link.message_queue'];

        if (! isset($config['db'])) {
            throw new \RuntimeException('Missing db settings for prooph.link.message_queue in application config');
        }

        if ($config['db']['use_application_db']) {
            $connection = $serviceLocator->get('prooph.link.app.db');
        } else {
            if (! isset($config['db']['connection'])) {
                throw new \RuntimeException('Missing db connection settings for prooph.link.message_queue in application config');
            }

            $connection = DriverManager::getConnection($config['db']['connection']);
        }

        return new PersistentFactory(new DoctrineDriver($connection), new BernardSerializer());
    }
}